File: scala-server/app/com/bolour/eagerwords/common/message/MachinePlayResponse.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
 */
package com.bolour.eagerwords.common.message

import com.bolour.eagerwords.common.domain.{GameMiniState, PlayPiece}
import com.bolour.plane.domain.Point

/**
  * API response to a request to initiate a machine play.
  *
  * @param gameMiniState Mini-state of the game after the play.
  * @param playedPieces List of pieces and locations used in the machine play.
  * @param deadPoints Empty points on the board that have been determined to be
  *                   no longer playable after the machine's play was laid out on the board.
  */
case class MachinePlayResponse(
  gameMiniState: GameMiniState,
  playedPieces: List[PlayPiece],
  deadPoints: List[Point]
)

File: haskell-server/src/EagerWords/Common/Message/MachinePlayResponse.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DeriveAnyClass #-}

module EagerWords.Common.Message.MachinePlayResponse (
    MachinePlayResponse(..)
  , tupleToMachinePlayResponse
  ) where

import GHC.Generics (Generic)
import Data.Aeson (FromJSON, ToJSON)
import Control.DeepSeq (NFData)

import Bolour.Plane.Domain.Point
import EagerWords.Common.Domain.PlayPiece
import EagerWords.Common.Domain.GameMiniState

data MachinePlayResponse = MachinePlayResponse {
    gameMiniState :: GameMiniState
  , playedPieces :: [PlayPiece]
  , deadPoints :: [Point]
}
  deriving (Eq, Show, Generic, NFData)

instance FromJSON MachinePlayResponse
instance ToJSON MachinePlayResponse

tupleToMachinePlayResponse :: (GameMiniState, [PlayPiece], [Point]) -> MachinePlayResponse
tupleToMachinePlayResponse (miniState, playPieces, deadPoints) =
  MachinePlayResponse miniState playPieces deadPoints

